### ISSUE TITLE
account sh pu for /result calls

### ISSUE THREAD

**Issue by jdries** [Open-EO/openeo-geopyspark-driver#281]

**Comment by bossie** [comment]
Involves logging PUs in the ETL API for `/result` calls so they are visible in the [Marketplace dashboard](https://kibana-apps.vgt.vito.be/app/kibana#/dashboard/c141ce70-a75b-11eb-bc45-753cf0cde572?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-7d,to:now))&_a=(description:'',filters:!(('$state':(store:appState),exists:(field:orchestrator),meta:(alias:!n,disabled:!f,index:ccab6f80-a758-11eb-bc45-753cf0cde572,key:orchestrator,negate:!f,type:exists,value:exists)),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:ccab6f80-a758-11eb-bc45-753cf0cde572,key:status,negate:!t,params:(query:skipped),type:phrase),query:(match:(status:(query:skipped,type:phrase))))),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),panels:!((embeddableConfig:(),gridData:(h:18,i:'3b585e89-a403-4351-b320-0c666c8f695b',w:16,x:0,y:0),id:a3e8af10-a75b-11eb-bc45-753cf0cde572,panelIndex:'3b585e89-a403-4351-b320-0c666c8f695b',type:visualization,version:'7.5.1'),(embeddableConfig:(),gridData:(h:18,i:d2309e96-90f2-46ab-99d1-44f5f67897b4,w:15,x:16,y:0),id:'92b57ec0-1484-11ed-9c26-3f76720e6a27',panelIndex:d2309e96-90f2-46ab-99d1-44f5f67897b4,type:visualization,version:'7.5.1'),(embeddableConfig:(title:'Marketplace%20-%20Total%20credits',vis:(colors:(billed:%23508642,skipped:%23F2C96D))),gridData:(h:18,i:f72542a1-1627-486e-87f3-6bdb93ad6e1a,w:17,x:31,y:0),id:d724b8f0-a75c-11eb-bc45-753cf0cde572,panelIndex:f72542a1-1627-486e-87f3-6bdb93ad6e1a,title:'Marketplace%20-%20Total%20credits',type:visualization,version:'7.5.1'),(embeddableConfig:(),gridData:(h:16,i:'63dadab4-557a-4a02-b67a-e72abd530762',w:12,x:0,y:18),id:d2084660-c9ea-11eb-bc45-753cf0cde572,panelIndex:'63dadab4-557a-4a02-b67a-e72abd530762',type:visualization,version:'7.5.1'),(embeddableConfig:(),gridData:(h:16,i:d066514b-182a-4f47-a817-0b0a22bd6946,w:16,x:12,y:18),id:d30ec610-a75d-11eb-bc45-753cf0cde572,panelIndex:d066514b-182a-4f47-a817-0b0a22bd6946,type:visualization,version:'7.5.1'),(embeddableConfig:(),gridData:(h:16,i:'9047a98b-59b6-405c-9566-edf02ed37679',w:20,x:28,y:18),id:'4c9030c0-a75c-11eb-bc45-753cf0cde572',panelIndex:'9047a98b-59b6-405c-9566-edf02ed37679',type:visualization,version:'7.5.1'),(embeddableConfig:(),gridData:(h:15,i:'55408dd5-c495-41fb-9710-3ee9087abfe9',w:16,x:0,y:34),id:'70bb6170-b6e0-11eb-bc45-753cf0cde572',panelIndex:'55408dd5-c495-41fb-9710-3ee9087abfe9',type:visualization,version:'7.5.1'),(embeddableConfig:(),gridData:(h:15,i:ad7ca16c-9773-4418-9f33-dfc297554f46,w:15,x:16,y:34),id:'783a0550-b6e0-11eb-bc45-753cf0cde572',panelIndex:ad7ca16c-9773-4418-9f33-dfc297554f46,type:visualization,version:'7.5.1'),(embeddableConfig:(),gridData:(h:15,i:'61beff2e-fae0-43e0-9159-76df7675a1ea',w:17,x:31,y:34),id:'744bf160-3025-11ec-9c26-3f76720e6a27',panelIndex:'61beff2e-fae0-43e0-9159-76df7675a1ea',type:visualization,version:'7.5.1'),(embeddableConfig:(),gridData:(h:15,i:'7501324f-2034-4017-8d5e-1b0cc0b703ec',w:16,x:0,y:49),id:'5e2c4c30-1f8b-11ed-9c26-3f76720e6a27',panelIndex:'7501324f-2034-4017-8d5e-1b0cc0b703ec',type:visualization,version:'7.5.1'),(embeddableConfig:(),gridData:(h:15,i:aea4ba94-dbad-4c04-aac6-7506f0a5d93f,w:15,x:16,y:49),id:c5f30c20-c2d3-11eb-bc45-753cf0cde572,panelIndex:aea4ba94-dbad-4c04-aac6-7506f0a5d93f,type:visualization,version:'7.5.1'),(embeddableConfig:(),gridData:(h:15,i:'1c2fe732-e901-4691-b5a2-67e9688c1a0f',w:17,x:31,y:49),id:'841fbda0-a75a-11eb-bc45-753cf0cde572',panelIndex:'1c2fe732-e901-4691-b5a2-67e9688c1a0f',type:visualization,version:'7.5.1')),query:(language:kuery,query:''),timeRestore:!f,title:'Marketplace%20-%20Costs',viewMode:view)).

Would involve an accumulator like we have for batch jobs but limited to the scope of this request. We don't want to do this in all of our environments (similar to #325); maybe a `SparkListener` is an option.

Returning the [`OpenEO-Costs` header](https://openeo.org/documentation/1.0/developers/api/reference.html#tag/Data-Processing/operation/compute-result) is not in scope.

**Comment by bossie** [comment]
This is odd. Sometimes executors try to add PUs after the Spark work has completed and the driver has already reported the total to the ETL API; FWIW this total still seems to be correct.

![additional_add_pus_calls](https://user-images.githubusercontent.com/1032518/233006161-f1af9b68-871f-4861-a659-407ad36f226a.jpg)

At first I thought this was a logging issue where logs from one `/result` call might accidentally carry the `req_id` from another `/result` call if they run in parallel (i.e. thread pinning doesn't work); however, this `req_id` (r-a3c5b279f232492c8e567ab506e98859) is also visible in the log messages themselves and in this case, it is propagated through regular Spark serialization from driver to executors so thread pinning does not apply:

> ScopedMetadataTracker(r-a3c5b279f232492c8e567ab506e98859) added 0.3333333432674408 PUs

I enabled debug logging of SHub `/process` calls to investigate further.

**Comment by bossie** [comment]
So with speculation enabled, the accumulator's total seems to be OK wrt/ billing the user (don't pay for requests sent because of speculation) but not for tracking the actual number of calls towards SHub.

**Comment by bossie** [comment]
@jdries not sure how to proceed. I gather that speculative execution is a means to temporarily fix performance bottlenecks but since it increases the amount of calls to SHub and reduces the effectivity of of accumulators to track these calls, might it make sense to disable it?

**Comment by jdries** [comment]
My ideal solution would be to only disable it for stages that load data from sentinelhub, but I couldn't immediately find a way to do that.
We can also make it less aggressive if we notice that speculation happens too fast?

**Comment by jdries** [comment]
@bossie  I just read that spark 3.4.0, released very recently, did a lot with respect to speculation.
My suggestion would be to close this one, and create a new one for the spark upgrade. This upgrade issue can then point to this, as a motivation for the upgrade.
https://issues.apache.org/jira/browse/SPARK-32170

**Comment by bossie** [comment]
Closing after enabling ETL API auth on CreoDIAS (but there are no SHub collections on there ([yet](https://github.com/eu-cdse/openeo-cdse-infra/issues/2)) to verify).
